File: frontend/src/services/subject.ts
import api from "./api";

export interface Subject {
  id: number;
  semester_id: number;
  name: string;
  code?: string;
  faculty?: string;
  min_attendance_percent: number;
  units_per_class: number;
  units_earned_per_class: number;
  units_lost_per_class: number;
  track_attendance: boolean;
  active_from?: string;
  active_until?: string;
}

export const subjectService = {
  async list(semesterId: number, inTimetableOnly?: boolean): Promise<Subject[]> {
    const response = await api.get(`/semesters/${semesterId}/subjects`, {
      params: inTimetableOnly ? { in_timetable_only: true } : undefined,
    });
    return response.data;
  },


  async create(semesterId: number, data: { name: string; code?: string; faculty?: string; min_attendance_percent: number; units_per_class?: number; units_earned_per_class?: number; units_lost_per_class?: number; track_attendance?: boolean; active_from?: string; active_until?: string }): Promise<Subject> {
    const response = await api.post(`/semesters/${semesterId}/subjects`, data);
    return response.data;
  },

  async update(semesterId: number, subjectId: number, data: Partial<Subject>): Promise<Subject> {
    const response = await api.put(`/semesters/${semesterId}/subjects/${subjectId}`, data);
    return response.data;
  },

  async syncAttendance(semesterId: number, subjectId: number, data: { attended: number; missed: number; delivered: number }): Promise<Subject> {
    const response = await api.put(`/semesters/${semesterId}/subjects/${subjectId}/sync-attendance`, data);
    return response.data;
  },

  async delete(semesterId: number, subjectId: number): Promise<void> {
    await api.delete(`/semesters/${semesterId}/subjects/${subjectId}`);
  },
};

File: backend/app/schemas/subject.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import date

class SubjectBase(BaseModel):
    name: str
    code: Optional[str] = None
    faculty: Optional[str] = None
    min_attendance_percent: float = Field(default=75.0, ge=0.0, le=100.0)
    units_per_class: int = Field(default=1, ge=1)
    units_earned_per_class: int = Field(default=1, ge=1)
    units_lost_per_class: int = Field(default=1, ge=1)
    track_attendance: bool = Field(default=True)
    active_from: Optional[date] = None
    active_until: Optional[date] = None

class SubjectCreate(SubjectBase):
    track_attendance: Optional[bool] = None

class SubjectUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None
    faculty: Optional[str] = None
    min_attendance_percent: Optional[float] = Field(default=None, ge=0.0, le=100.0)
    units_per_class: Optional[int] = Field(default=None, ge=1)
    units_earned_per_class: Optional[int] = Field(default=None, ge=1)
    units_lost_per_class: Optional[int] = Field(default=None, ge=1)
    track_attendance: Optional[bool] = None
    active_from: Optional[date] = None
    active_until: Optional[date] = None

class SubjectResponse(SubjectBase):
    id: int
    semester_id: int

    class Config:
        from_attributes = True

class SubjectAttendanceSyncRequest(BaseModel):
    attended: int = Field(..., ge=0)
    missed: int = Field(..., ge=0)
    delivered: int = Field(..., ge=0)
